src/transform/src/monotonic.rs

```diff
@@ -120,38 +123,84 @@ impl MonotonicFlag {
                 MirRelationExpr::LetRec {
                     ids,
                     values,
-                    max_iters: _,
+                    max_iters,
                     body,
                 } => {
+                    // Optimistically assume that all bindings were monotonic
+                    // (it is safe to do this as we initialize each binding to
+                    // the empty collection, which is trivially monotonic).
                     for id in ids.iter() {
-                        if locals.contains(id) {
-                            panic!("Shadowing of identifier: {:?}", id);
-                        }
+                        let inserted = locals.insert(id.clone());
+                        assert!(inserted, "Shadowing of identifier: {id:?}");
                     }
-                    // Pessimistically assume all bindings are non-monotonic, and
-                    // iteratively attempt to determine that some may be monotonic.
+
+                    // Sequentially AND locals[id] with the result of descending
```

Comment:
Maybe add a note in the comment that the lattice here looks as follows (it's a bit counterintuitive):
- bottom is true / "present in locals"
- top is false / "not present in locals"
- join is AND
- (meet would be OR, but this is not used here)